File: R/gauss_exponential.R
# This code illustrates how to fit a model to observations
# using the standard chi^2 minimization technique.
# WARNING: Here we simulate a counting experiment (poisson distribution)
#          and thus the chi^2 technique is only valid asymptotically.
#          For low count data one would use a maximum likelihood technique instead.
#          This code is not optimized for speed and is meant for educational purpose only. 

library(Hmisc) # provides errorbar function "errbar"
rm(list=ls()) # clear memory
setwd("/Users/hans/Analysis/teaching/PHY505/fitting/")

# define the model (gaussian with exponentially decaying background)
model <- function(x, pars) {
  A=pars[1]; B=pars[2]; C=pars[3]; D=pars[4]; E=pars[5]
  A*exp(-B*x)+C*exp(-1/2*((x-D)/E)^2)
}

# parameters
a=1000 # exp. norm.
b=0.2 # exp. slope
c=100 # gauss norm
mu=8 # peak location
sigma=1 #1 # width
seed = c(a, b, c, mu, sigma)

# create model points for plotting
x<-seq(0,20,0.01)
y<-model(x, seed)

pts_model = data.frame(
  x  = x
  , y  = y
)

# create model points on sampling grid (bins)
x<-seq(1,19,by=1)
y<-model(x, seed)

# generate random numbers from poisson distribution in each "bin"
set.seed(1, kind = NULL, normal.kind = NULL) # change seed to obtained new dataset
yobs = rpois(length(y), y)
pts_obs = data.frame(
  x  = x
  , y  = yobs
  , sd = sqrt(yobs)
)

# define chi2 function
chi2 <- function(pars) {
    yobs = pts_obs$y
    xobs = pts_obs$x
    yexp = model(xobs, pars)
    sum((yobs-yexp)^2/yexp)
    # alternatively use sum((yobs-yexp)^2/yobs)
}

# do a numerical minimization
result = optim(seed, chi2, gr = NULL, method = "BFGS")
x = seq(0,20,by=0.01)
y = model(x, result$par)

pts_fit = data.frame(
  x  = x
  , y  = y
)

# plot the result
png(file="fitresult.png",width=1500,height=1250, res=250)
# line of original model
with (
  data = pts_model
  , expr = plot(x, y, "l", col="red", xlab="energy [keV]", ylab="counts", ylim=c(0.0,800.0), las=1)
)
# line of best fit
with (
  data = pts_fit
  , expr = lines(x, y, col="green", lwd=2)
)
# data points ...
with (
  data = pts_obs
  , expr = points(x, y, col="black", cex = .5,pch=21,bg="black", add=T)
)
# ... and errorbars
with (
  data = pts_obs
  , expr = errbar(x, y, y+sd, y-sd, add=T, pch=1, cap=.01, cex = .5)
)
legend(8,800, c("true model","fit result"), lty=c(1,1), lwd=c(2.5,2.5),col=c("red","green")) 
dev.off()

# perform crude uncertainty calculation 
# crude == ignore possible correlations between fit parameters (can be drastic underestimate)
# use mean of gaussian as example

reduced_model <- function(x, mean) {
   model(x, c(result$par[1], result$par[2], result$par[3], mean, result$par[5]))
}

delta_chi2 <- function(mean) {
  yobs = pts_obs$y
  xobs = pts_obs$x
  yexp = reduced_model(xobs, mean)
  # define chi2_min = 0
  sum((yobs-yexp)^2/yexp)-chi2(result$par)
}

# visualize - ranges are hardcoded
scan_x = seq(7.4,8.4, by=0.01)
scan_y = sapply(scan_x, delta_chi2)

png(file="chi2_mean.png",width=1500,height=1250, res=250)
plot(scan_x, scan_y, "l", col="black", xlab="", ylab="", las=1)
mtext(expression(paste(Delta, chi^2)), 2, 2.0, cex = 1.5)
mtext("mean [kev]", 1, 3.0, cex = 1.5)
dev.off()

# solve for Delta Chi2 == 1 via root finding.
f <- function (x, a) delta_chi2(x) - a
x1 = uniroot(f, c(result$par[4]-0.3, result$par[4]), tol = 0.0001, a = 1)
x2 = uniroot(f, c(result$par[4], result$par[4]+0.3), tol = 0.0001, a = 1)

paste("the mean is: ", result$par[4], " + ", x2$root-result$par[4], " - ", result$par[4]-x1$root)

# calculate uncertainties that take into account correlation:
# need to re-minimize nuisance parameters 
# (everything that is not the mean in our example)
# for each value of the parameter of interest 
# (in our case the mean)
# we need a nested function definition to this

minimize_stepwise <- function(mean) {
  reduced_model <- function(x, pars) {
    model(x, c(pars[1], pars[2], pars[3], mean, pars[4]))
  }
  delta_chi2 <- function(pars) {
    yobs = pts_obs$y
    xobs = pts_obs$x
    yexp = reduced_model(xobs, pars)
    # define chi2_min = 0
    sum((yobs-yexp)^2/yexp)-chi2(result$par)
  }
  seed = c(seed[1], seed[2], seed[3], seed[5])
  result2 = optim(seed, delta_chi2, gr = NULL, method = "BFGS")
  result2$value
}

scan_y = sapply(scan_x, minimize_stepwise)

png(file="chi2_mean_profile.png",width=1500,height=1250, res=250)
plot(scan_x, scan_y, "l", col="black", xlab="", ylab="", las=1)
mtext(expression(paste(Delta, chi^2)), 2, 2.0, cex = 1.5)
mtext("mean [kev]", 1, 3.0, cex = 1.5)
dev.off()

# solve for Delta Chi2 == 1 via root finding.
f <- function (x, a) minimize_stepwise(x) - a
x1 = uniroot(f, c(result$par[4]-0.3, result$par[4]), tol = 0.0001, a = 1)
x2 = uniroot(f, c(result$par[4], result$par[4]+0.3), tol = 0.0001, a = 1)

paste("the mean is: ", result$par[4], " + ", x2$root-result$par[4], " - ", result$par[4]-x1$root)

File: PYTHON/gauss_exponential.py
#!/usr/bin/env python

# This code illustrates how to fit a model to observations
# using the standard chi^2 minimization technique.
# WARNING: Here we simulate a counting experiment (poisson distribution)
#          and thus the chi^2 technique is only valid asymptotically.
#          For low count data one would use a maximum likelihood technique instead.
#          This code is not optimized for speed and is meant for educational purpose only. 

import numpy
import pylab as plt
from scipy import optimize

def model(x, pars):
	A=pars[0];
	B=pars[1];
	C=pars[2];
	mean=pars[3];
	sigma=pars[4];
	return A*numpy.exp(-B*x)+C*numpy.exp(-1./2.*((x-mean)/sigma)**2)

# parameters
a=1000 # exp. norm.
b=0.2 # exp. slope
c=100 # gauss norm
mean=8 # peak location
sigma=1 # width
seed = [a, b, c, mean, sigma]

x = numpy.arange(0, 20, 0.01)
y = model(x, seed)

x_pts = numpy.arange(1., 20., 1.0)
y_exp = model(x_pts, seed)
#y_obs = numpy.random.poisson(y_exp)
y_obs = [800, 704, 578, 458, 339, 323, 320, 311, 221, 139, 105, 87, 73, 60, 56, 46, 36, 32, 26]
y_errs = numpy.sqrt(y_obs)

chi2 = lambda pars: numpy.sum((y_obs-model(x_pts,pars))**2/model(x_pts,pars))
result = optimize.minimize(chi2, seed, method='Nelder-Mead')
chi2_min = result.fun
print "best fit is:", result.x, "with chi2:", chi2_min

y_fit = model(x, result.x)

fig = plt.figure(figsize=(7, 6), dpi=200)
ax = plt.axes()
ax.plot(x,y,"b-",label="true model")
ax.plot(x,y_fit,"r-",label="best fit")
ax.errorbar(x_pts, y_obs, yerr=y_errs, xerr=False, axes=ax, fmt='ko', markersize=2.0)
ax.set_ylabel('counts', position=(0., 1.), va='top', ha='right')
ax.set_xlabel('energy [keV]', position=(1., 0.), va='bottom', ha='right')
ax.yaxis.set_label_coords(-0.12, 1.)
ax.xaxis.set_label_coords(1.0, -0.1)
ax.set_ylim([0, 1.e3])
ax.legend()
plt.savefig("fit.png")

# calculate uncertainties for mean
means = numpy.linspace(result.x[3]-0.3, result.x[3]+0.3, 100).tolist()
chi2vals = []
for mu in means:
       	def chi2_rel(pars, mu=mu):		
    		current_pars = list(pars[:3])+[mu]+list(pars[3:])		
    		y_exp = model(x_pts, current_pars)
    		return numpy.sum((y_obs-y_exp)**2/y_exp) - chi2_min
    
    	x0 = [seed[0], seed[1], seed[2], seed[4]]
    	res = optimize.minimize(chi2_rel, x0, method='Nelder-Mead')
	chi2vals.append(res.fun)

fig = plt.figure(figsize=(7, 6), dpi=200)
ax = plt.axes()
ax.plot(means, chi2vals, "k-")
ax.set_ylabel('$\\Delta\\chi^2$', position=(0., 1.), va='top', ha='right')
ax.set_xlabel('mean [keV]', position=(1., 0.), va='bottom', ha='right')
ax.yaxis.set_label_coords(-0.12, 1.)
ax.xaxis.set_label_coords(1.0, -0.1)
plt.savefig("chi2.png")

# find roots numerically via interpolation
from scipy.interpolate import InterpolatedUnivariateSpline
spline = InterpolatedUnivariateSpline(means, numpy.asarray(chi2vals)-1.)
roots = spline.roots()
print "best fit:", result.x[3], "+", roots[1]-result.x[3], "-", result.x[3]-roots[0]

 
    

      






